### PR TITLE
NPC Auto Run

### PR DESCRIPTION
**Pull request by Evanosity**
Two new fields for NPCDefs(probably curves)
-autoRunThreshold, 0-100; the % hp at which this monster will begin attempting to run. Default 0.
-autoRunChance, 0-100; the chance that this monster will attempt to run. Default 0.

When the player attacks, before damage calculation, it checks autoRunThreshold, then autoRunChance, then the monster has to pass a dexterity check. If all 3 pass, the monster gets deleted and the attacker leaves combat.

### CODE REVIEW

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 1]
```diff
@@ -4491,12 +4493,43 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
+        
+        String status = "";
+        
+        int autoRunThreshold = targetCharacter.getProperty("autoRunThreshold");
+        if(autoRunThreshold == null) autoRunThreshold = 0;
+        int autoRunChance = targetCharacter.getProperty("autoRunChance");
+        if(autoRunChance == null) autoRunChance = 0;
+        
+        //First, we check against the auto run threshold to see if the monster will attempt to run.
+        if((targetCharacter.getProperty("hitpoints") / targetCharacter.getProperty("maxHitpoints"))*100 < autoRunThreshold) {
```
Comment (Emperorlou) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4505]: getProperty("hitpoints") is not cast to a number so you would have gotten compiler errors here. You can't submit a PR if you have compiler errors so I'll stop looking at this until you bring it up and make sure compiler errors are taken care of first.

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 2]
```diff
@@ -4491,12 +4493,46 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
+        if(autoRunThreshold == null) autoRunThreshold = 0;
+        Integer autoRunChance = (Integer) targetCharacter.getProperty("autoRunChance");
+        if(autoRunChance == null) autoRunChance = 0;
+        
+        int targetHP = (int) targetCharacter.getProperty("hitpoints");
+        int targetMaxHP = (int) targetCharacter.getProperty("maxHitpoints");
+        
+        //First, we check against the auto run threshold to see if the monster will attempt to run.
+        if((targetHP / targetMaxHP)*100 < autoRunThreshold) {
+
```
Comment (SPFiredrake) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4509]: `int / int` = `int`. You'll always get a 0 here, unless you cast to double. Do `100d * targetHP / targetMaxHP` instead.

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 3]
```diff
@@ -4491,12 +4493,46 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
+        if(autoRunChance == null) autoRunChance = 0;
+        
+        int targetHP = (int) targetCharacter.getProperty("hitpoints");
+        int targetMaxHP = (int) targetCharacter.getProperty("maxHitpoints");
+        
+        //First, we check against the auto run threshold to see if the monster will attempt to run.
+        if((targetHP / targetMaxHP)*100 < autoRunThreshold) {
+            
+        	//then, we check a random number vs the auto run chance.
+        	if (rnd.nextDouble()*100 <= autoRunChance) {
```
Comment (SPFiredrake) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4511]: Use `GameUtils.roll` here.

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 4]
```diff
@@ -4491,12 +4493,46 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
+            	
+            	//then, the monster has to pass a dexterity check.
+            	if(rnd.nextDouble() * charDex <= rnd.nextDouble() * monsterDex) {
+            		cs.leaveCombat(sourceCharacter, targetCharacter);
+            		
+            		db.put(sourceCharacter);
+            		db.delete(targetCharacter);
```
Comment (SPFiredrake) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4518]: Make sure NPC items get deleted too.

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 5]
```diff
@@ -4491,12 +4493,46 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
         Double monsterDex = getCharacterDexterity(targetCharacter);
         Random rnd = new Random();
+        
+        String status = "";
+        
+        Integer autoRunThreshold = (Integer) targetCharacter.getProperty("autoRunThreshold");
+        if(autoRunThreshold == null) autoRunThreshold = 0;
+        Integer autoRunChance = (Integer) targetCharacter.getProperty("autoRunChance");
+        if(autoRunChance == null) autoRunChance = 0;
+
```
Comment (SPFiredrake) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4503]: Only run this logic if both `autoRunThreshold > 0 && autoRunChance > 0`. Spare some CPU cycles.

File: Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java [hunk 6]
```diff
@@ -4491,12 +4493,46 @@ public String doCharacterAttemptAttack(ODPAuthenticator auth, CachedEntity user,
...
+        
+        String status = "";
+        
+        Integer autoRunThreshold = (Integer) targetCharacter.getProperty("autoRunThreshold");
+        if(autoRunThreshold == null) autoRunThreshold = 0;
+        Integer autoRunChance = (Integer) targetCharacter.getProperty("autoRunChance");
+        if(autoRunChance == null) autoRunChance = 0;
+        
+        int targetHP = (int) targetCharacter.getProperty("hitpoints");
+        int targetMaxHP = (int) targetCharacter.getProperty("maxHitpoints");
```
Comment (SPFiredrake) [inline Initium-Odp/src/com/universeprojects/miniup/server/ODPDBAccess.java:4505]: These are both `Double` type.

**Pull request merged**